### PR TITLE
remove CEL validation, validate inside the webhook

### PR DESCRIPTION
**Pull request by charles-chenzz**
try to solve #60 

but might be wrong

### CODE REVIEW

**Review by tenzen-y — COMMENTED** [review]
Hello, @charles-chenzz.

Probably, you need to implement validations to Validator instead of Defaulter:

https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/api/v1alpha1/jobset_webhook.go#L48-L61

**Comment by charles-chenzz** [comment]
> Hello, @charles-chenzz.
> 
> Probably, you need to implement validations to Validator instead of Defaulter:
> 
> https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/api/v1alpha1/jobset_webhook.go#L48-L61

will take a look at it.

**Comment by ahg-g** [comment]
Correct, we need to"

1) add the validation to the validation webhook
2) remove [this integration test](https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/test/integration/controller/jobset_controller_test.go#L122)
3) add integration tests to [this table](https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/test/integration/webhook/jobset_webhook_test.go#L72)

**Comment by charles-chenzz** [comment]
> Correct, we need to"
> 
> 1. add the validation to the validation webhook
> 2. remove [this integration test](https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/test/integration/controller/jobset_controller_test.go#L122)
> 3. add integration tests to [this table](https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/test/integration/webhook/jobset_webhook_test.go#L72)

Regarding the first point, I need to move the logic that I add in the defaulter to the ``ValidateCreate()`` / ``ValidateUpdate()`` / ``ValidateDelete()`` ? after I done for that, remove the [test](https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/test/integration/controller/jobset_controller_test.go#L122) .

**Comment by tenzen-y** [comment]
> > Correct, we need to"
> > 
> > 1. add the validation to the validation webhook
> > 2. remove [this integration test](https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/test/integration/controller/jobset_controller_test.go#L122)
> > 3. add integration tests to [this table](https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/test/integration/webhook/jobset_webhook_test.go#L72)
> 
> Regarding the first point, I need to move the logic that I add in the defaulter to the `ValidateCreate()` / `ValidateUpdate()` / `ValidateDelete()` ? after I done for that, remove the [test](https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/test/integration/controller/jobset_controller_test.go#L122) .

I think it is enough to add validations only to the ValidateCreate() / ValidateUpdate().

**Comment by ahg-g** [comment]
Correct, you actually just need to add it on ValidateCreate since those fields are immutable

**Comment by charles-chenzz** [comment]
> add integration tests to [this table](https://github.com/kubernetes-sigs/jobset/blob/f62e66440949f04582934c4623c49d199cf3734a/test/integration/webhook/jobset_webhook_test.go#L72)

working on this now

File: api/v1alpha1/jobset_webhook.go [hunk 1]
```diff
@@ -56,6 +57,14 @@ var _ webhook.Validator = &JobSet{}
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type
 func (r *JobSet) ValidateCreate() error {
+	for i := range r.Spec.ReplicatedJobs {
+		if r.Spec.ReplicatedJobs[i].Network == nil || !*r.Spec.ReplicatedJobs[i].Network.EnableDNSHostnames ||
+			(r.Spec.ReplicatedJobs[i].Template.Spec.CompletionMode != nil &&
+				*r.Spec.ReplicatedJobs[i].Template.Spec.CompletionMode == batchv1.IndexedCompletion) {
+			// how do we record the error message here?
+			fmt.Println("EnableDNSHostnames requires job to be in indexed completion mode")
```
Comment (malwaredllc) [inline api/v1alpha1/jobset_webhook.go:65]: I think you can just do `return  fmt.Errorf("EnableDNSHostnames requires job to be in indexed completion mode")`

Comment (danielvegamyhre) [inline api/v1alpha1/jobset_webhook.go:65]:
```suggestion
return fmt.Errorf("EnableDNSHostnames requires job to be in indexed completion mode")
```

Comment (charles-chenzz) [reply api/v1alpha1/jobset_webhook.go:65]: will change it later

Comment (charles-chenzz) [reply api/v1alpha1/jobset_webhook.go]: done

File: api/v1alpha1/jobset_webhook.go [hunk 2]
```diff
@@ -56,6 +57,13 @@ var _ webhook.Validator = &JobSet{}
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type
 func (r *JobSet) ValidateCreate() error {
+	for i := range r.Spec.ReplicatedJobs {
+		if r.Spec.ReplicatedJobs[i].Network == nil || !*r.Spec.ReplicatedJobs[i].Network.EnableDNSHostnames ||
```
Comment (ahg-g) [inline api/v1alpha1/jobset_webhook.go:61]: do it like this:

```
enableDNSHostnames := .Spec.ReplicatedJobs[i].Network != nil &&
Spec.ReplicatedJobs[i].Network.EnableDNSHostnames != nil &&  *r.Spec.ReplicatedJobs[i].Network.EnableDNSHostnames
indexed := r.Spec.ReplicatedJobs[i].Template.Spec.CompletionMode != nil &&
				*r.Spec.ReplicatedJobs[i].Template.Spec.CompletionMode == batchv1.IndexedCompletion

if enableDNSHostnames && !indexed {
  return fmt.Errorf("EnableDNSHostnames requires job to be in Indexed completion mode")
}

```

Comment (charles-chenzz) [reply api/v1alpha1/jobset_webhook.go]: done fot it. In fact, I want to assign a new var.
like this ``replicatedJobs := r.Spec.ReplicatedJobs`` and the rest of ``r.Spec.ReplicatedJobs[i]`` can change to ``replicatedJobs[i]``. what do you think?

File: test/integration/webhook/jobset_webhook_test.go [hunk 3]
```diff
@@ -130,5 +130,16 @@ var _ = ginkgo.Describe("jobset webhook defaulting", func() {
 				return js.Spec.ReplicatedJobs[0].Network != nil && *js.Spec.ReplicatedJobs[0].Network.EnableDNSHostnames
 			},
 		}),
+		ginkgo.Entry("validate enableDNSHostnames can't be set if job is not Indexed", &testCase{
```
Comment (charles-chenzz) [inline test/integration/webhook/jobset_webhook_test.go:133]: this place is the reason why test failed, however I'm not familiar with integration test and this testcase might have error.. I have no clues now..

Comment (danielvegamyhre) [reply test/integration/webhook/jobset_webhook_test.go:153]: all tests are passing now so looks like you figured it out

**Comment by danielvegamyhre** [comment]
@ahg-g this LGTM do you want to take a look and maybe we can merge this today before 0.1.0 release?

**Pull request merged**